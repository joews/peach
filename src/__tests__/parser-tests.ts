/* eslint-env jest */
import parse from '../parser'
import { fixture } from './helpers'

//
// snapshot tests for the parser
//

function testFixture (fixtureName: string) {
  test(fixtureName, () => {
    expect(parse(fixture(fixtureName))).toMatchSnapshot()
  })
}

function testParse (source: string) {
  test(source, () => {
    // console.log(JSON.stringify(parse(source)))
    expect(parse(source)).toMatchSnapshot()
  })
}

// TODO more, better organised grammar tests!
testFixture('fibonacci.peach')
testFixture('str.peach')
testFixture('tail-recursion.peach')

testParse('(true => 1)')
testParse('([true|tail] => 1)')
testParse('((1, 2) => 1)()')

testParse(`[]`)
testParse(`[1,2,3]`)

testParse(`(test)`)
testParse(`test(1)`)
testParse(`(test (1, 2))`)

testParse(`t()`)
testParse(`t(1, 2)`)
testParse(`t(\`1\`, (2), (x) => 3)`)

testParse(`1`)
testParse(`1 * 2/( 2 % 3)`)
testParse(`1 * 2 + 2 - 3`)
testParse(`1 * 2 + (bob - [3])`)
testParse(`2 * (1 + a < b) < 3`)
testParse('a = `two`')
testParse(`a = b < 2`)
testParse(`a == 2`)
testParse(`a == b <=> true`)

testParse(`a.b`)
testParse(`a.b.c`)
testParse(`a[b][0]`)
testParse(`a.b[0]`)

testParse(`a()`)
testParse(`a(b)`)
testParse(`a(b)(c)`)

testParse(`a => 1`)
testParse(`(a) => 1`)
testParse(`(a, b) => {
  1
  2
}`)

testParse(`(a, b) => {
  1
  2
}, (c, d) => true`)

testParse(`
a =
  (1) => a,
  (1) => 2
`)

testParse(`if (a) 1 else 2`)
testParse(`a < if (a) 1 else 2`)
testParse(`()`)
testParse(`(1,)`)
testParse(`(1,2)`)
testParse(`(1, 2, 3,)`)
testParse(`(1, (1 + 2, ((3,),)), (), (2), (3,))`)
testParse(`((((()))))`)
