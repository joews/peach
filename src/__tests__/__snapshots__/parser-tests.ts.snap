// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`
a =
  (1) => a,
  (1) => 2
 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "Def",
      "name": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "value": Object {
        "clauses": Array [
          Object {
            "body": Array [
              Object {
                "kind": "Identifier",
                "name": "a",
              },
            ],
            "pattern": Array [
              Object {
                "kind": "Number",
                "value": 1,
              },
            ],
          },
          Object {
            "body": Array [
              Object {
                "kind": "Number",
                "value": 2,
              },
            ],
            "pattern": Array [
              Object {
                "kind": "Number",
                "value": 1,
              },
            ],
          },
        ],
        "kind": "Function",
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`(a) => 1 1`] = `
Object {
  "expressions": Array [
    Object {
      "clauses": Array [
        Object {
          "body": Array [
            Object {
              "kind": "Number",
              "value": 1,
            },
          ],
          "pattern": Array [
            Object {
              "kind": "Identifier",
              "name": "a",
            },
          ],
        },
      ],
      "kind": "Function",
    },
  ],
  "kind": "Program",
}
`;

exports[`(a, b) => {
  1
  2
} 1`] = `
Object {
  "expressions": Array [
    Object {
      "clauses": Array [
        Object {
          "body": Array [
            Object {
              "kind": "Number",
              "value": 1,
            },
            Object {
              "kind": "Number",
              "value": 2,
            },
          ],
          "pattern": Array [
            Object {
              "kind": "Identifier",
              "name": "a",
            },
            Object {
              "kind": "Identifier",
              "name": "b",
            },
          ],
        },
      ],
      "kind": "Function",
    },
  ],
  "kind": "Program",
}
`;

exports[`(a, b) => {
  1
  2
}, (c, d) => true 1`] = `
Object {
  "expressions": Array [
    Object {
      "clauses": Array [
        Object {
          "body": Array [
            Object {
              "kind": "Number",
              "value": 1,
            },
            Object {
              "kind": "Number",
              "value": 2,
            },
          ],
          "pattern": Array [
            Object {
              "kind": "Identifier",
              "name": "a",
            },
            Object {
              "kind": "Identifier",
              "name": "b",
            },
          ],
        },
        Object {
          "body": Array [
            Object {
              "kind": "Boolean",
              "value": true,
            },
          ],
          "pattern": Array [
            Object {
              "kind": "Identifier",
              "name": "c",
            },
            Object {
              "kind": "Identifier",
              "name": "d",
            },
          ],
        },
      ],
      "kind": "Function",
    },
  ],
  "kind": "Program",
}
`;

exports[`1 * 2 + (bob - [3]) 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "BinaryOperator",
      "left": Object {
        "kind": "BinaryOperator",
        "left": Object {
          "kind": "Number",
          "value": 1,
        },
        "operator": "*",
        "right": Object {
          "kind": "Number",
          "value": 2,
        },
      },
      "operator": "+",
      "right": Object {
        "kind": "BinaryOperator",
        "left": Object {
          "kind": "Identifier",
          "name": "bob",
        },
        "operator": "-",
        "right": Object {
          "kind": "Array",
          "values": Array [
            Object {
              "kind": "Number",
              "value": 3,
            },
          ],
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`1 * 2 + 2 - 3 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "BinaryOperator",
      "left": Object {
        "kind": "BinaryOperator",
        "left": Object {
          "kind": "Number",
          "value": 1,
        },
        "operator": "*",
        "right": Object {
          "kind": "Number",
          "value": 2,
        },
      },
      "operator": "+",
      "right": Object {
        "kind": "BinaryOperator",
        "left": Object {
          "kind": "Number",
          "value": 2,
        },
        "operator": "-",
        "right": Object {
          "kind": "Number",
          "value": 3,
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`1 * 2/( 2 % 3) 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "BinaryOperator",
      "left": Object {
        "kind": "Number",
        "value": 1,
      },
      "operator": "*",
      "right": Object {
        "kind": "BinaryOperator",
        "left": Object {
          "kind": "Number",
          "value": 2,
        },
        "operator": "/",
        "right": Object {
          "kind": "BinaryOperator",
          "left": Object {
            "kind": "Number",
            "value": 2,
          },
          "operator": "%",
          "right": Object {
            "kind": "Number",
            "value": 3,
          },
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`1 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "Number",
      "value": 1,
    },
  ],
  "kind": "Program",
}
`;

exports[`2 * (1 + a < b) < 3 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "BinaryOperator",
      "left": Object {
        "kind": "BinaryOperator",
        "left": Object {
          "kind": "Number",
          "value": 2,
        },
        "operator": "*",
        "right": Object {
          "kind": "BinaryOperator",
          "left": Object {
            "kind": "BinaryOperator",
            "left": Object {
              "kind": "Number",
              "value": 1,
            },
            "operator": "+",
            "right": Object {
              "kind": "Identifier",
              "name": "a",
            },
          },
          "operator": "<",
          "right": Object {
            "kind": "Identifier",
            "name": "b",
          },
        },
      },
      "operator": "<",
      "right": Object {
        "kind": "Number",
        "value": 3,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`a < if (a) 1 else 2 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "BinaryOperator",
      "left": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "operator": "<",
      "right": Object {
        "condition": Object {
          "kind": "Identifier",
          "name": "a",
        },
        "elseBranch": Object {
          "kind": "Number",
          "value": 2,
        },
        "ifBranch": Object {
          "kind": "Number",
          "value": 1,
        },
        "kind": "If",
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`a = \`two\` 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "Def",
      "name": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "value": Object {
        "kind": "String",
        "value": "two",
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`a = b < 2 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "Def",
      "name": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "value": Object {
        "kind": "BinaryOperator",
        "left": Object {
          "kind": "Identifier",
          "name": "b",
        },
        "operator": "<",
        "right": Object {
          "kind": "Number",
          "value": 2,
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`a == 2 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "BinaryOperator",
      "left": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "operator": "==",
      "right": Object {
        "kind": "Number",
        "value": 2,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`a == b <=> true 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "BinaryOperator",
      "left": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "operator": "==",
      "right": Object {
        "kind": "BinaryOperator",
        "left": Object {
          "kind": "Identifier",
          "name": "b",
        },
        "operator": "<=>",
        "right": Object {
          "kind": "Boolean",
          "value": true,
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`a => 1 1`] = `
Object {
  "expressions": Array [
    Object {
      "clauses": Array [
        Object {
          "body": Array [
            Object {
              "kind": "Number",
              "value": 1,
            },
          ],
          "pattern": Array [
            Object {
              "kind": "Identifier",
              "name": "a",
            },
          ],
        },
      ],
      "kind": "Function",
    },
  ],
  "kind": "Program",
}
`;

exports[`a() 1`] = `
Object {
  "expressions": Array [
    Object {
      "args": null,
      "fn": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "kind": "Call",
    },
  ],
  "kind": "Program",
}
`;

exports[`a(b) 1`] = `
Object {
  "expressions": Array [
    Object {
      "args": Array [
        Object {
          "kind": "Identifier",
          "name": "b",
        },
      ],
      "fn": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "kind": "Call",
    },
  ],
  "kind": "Program",
}
`;

exports[`a(b)(c) 1`] = `
Object {
  "expressions": Array [
    Object {
      "args": Array [
        Object {
          "kind": "Identifier",
          "name": "c",
        },
      ],
      "fn": Object {
        "args": Array [
          Object {
            "kind": "Identifier",
            "name": "b",
          },
        ],
        "fn": Object {
          "kind": "Identifier",
          "name": "a",
        },
        "kind": "Call",
      },
      "kind": "Call",
    },
  ],
  "kind": "Program",
}
`;

exports[`a.b 1`] = `
Object {
  "expressions": Array [
    Object {
      "computed": false,
      "kind": "Member",
      "name": Object {
        "kind": "Identifier",
        "name": "b",
      },
      "source": Object {
        "kind": "Identifier",
        "name": "a",
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`a.b.c 1`] = `
Object {
  "expressions": Array [
    Object {
      "computed": false,
      "kind": "Member",
      "name": Object {
        "kind": "Identifier",
        "name": "c",
      },
      "source": Object {
        "computed": false,
        "kind": "Member",
        "name": Object {
          "kind": "Identifier",
          "name": "b",
        },
        "source": Object {
          "kind": "Identifier",
          "name": "a",
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`a.b[0] 1`] = `
Object {
  "expressions": Array [
    Object {
      "computed": true,
      "kind": "Member",
      "name": Object {
        "kind": "Number",
        "value": 0,
      },
      "source": Object {
        "computed": false,
        "kind": "Member",
        "name": Object {
          "kind": "Identifier",
          "name": "b",
        },
        "source": Object {
          "kind": "Identifier",
          "name": "a",
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`a[b][0] 1`] = `
Object {
  "expressions": Array [
    Object {
      "computed": true,
      "kind": "Member",
      "name": Object {
        "kind": "Number",
        "value": 0,
      },
      "source": Object {
        "computed": true,
        "kind": "Member",
        "name": Object {
          "kind": "Identifier",
          "name": "b",
        },
        "source": Object {
          "kind": "Identifier",
          "name": "a",
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`if (a) 1 else 2 1`] = `
Object {
  "expressions": Array [
    Object {
      "condition": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "elseBranch": Object {
        "kind": "Number",
        "value": 2,
      },
      "ifBranch": Object {
        "kind": "Number",
        "value": 1,
      },
      "kind": "If",
    },
  ],
  "kind": "Program",
}
`;

exports[`tail-recursion.peach 1`] = `
Object {
  "expressions": Array [
    Object {
      "kind": "Def",
      "name": Object {
        "kind": "Identifier",
        "name": "factorial",
      },
      "value": Object {
        "clauses": Array [
          Object {
            "body": Array [
              Object {
                "kind": "Identifier",
                "name": "n",
              },
            ],
            "pattern": Array [
              Object {
                "kind": "Identifier",
                "name": "n",
              },
              Object {
                "kind": "Number",
                "value": 1,
              },
            ],
          },
          Object {
            "body": Array [
              Object {
                "args": Array [
                  Object {
                    "kind": "BinaryOperator",
                    "left": Object {
                      "kind": "Identifier",
                      "name": "n",
                    },
                    "operator": "*",
                    "right": Object {
                      "kind": "Identifier",
                      "name": "x",
                    },
                  },
                  Object {
                    "kind": "BinaryOperator",
                    "left": Object {
                      "kind": "Identifier",
                      "name": "x",
                    },
                    "operator": "-",
                    "right": Object {
                      "kind": "Number",
                      "value": 1,
                    },
                  },
                ],
                "fn": Object {
                  "kind": "Identifier",
                  "name": "factorial",
                },
                "kind": "Call",
              },
            ],
            "pattern": Array [
              Object {
                "kind": "Identifier",
                "name": "n",
              },
              Object {
                "kind": "Identifier",
                "name": "x",
              },
            ],
          },
        ],
        "kind": "Function",
      },
    },
    Object {
      "kind": "Def",
      "name": Object {
        "kind": "Identifier",
        "name": "f",
      },
      "value": Object {
        "clauses": Array [
          Object {
            "body": Array [
              Object {
                "args": Array [
                  Object {
                    "kind": "Number",
                    "value": 1,
                  },
                  Object {
                    "kind": "Identifier",
                    "name": "n",
                  },
                ],
                "fn": Object {
                  "kind": "Identifier",
                  "name": "factorial",
                },
                "kind": "Call",
              },
            ],
            "pattern": Array [
              Object {
                "kind": "Identifier",
                "name": "n",
              },
            ],
          },
        ],
        "kind": "Function",
      },
    },
    Object {
      "args": Array [
        Object {
          "kind": "Number",
          "value": 32768,
        },
      ],
      "fn": Object {
        "kind": "Identifier",
        "name": "f",
      },
      "kind": "Call",
    },
  ],
  "kind": "Program",
}
`;
